### ISSUE TITLE
Implement flake8-trio rules

### ISSUE THREAD

**Issue by charliermarsh** [astral-sh/ruff#8451]

**Comment by qdegraaf** [comment]
`TRIO118` sounds like it is not needed in Ruff, similar to `TRIO106`

The `TRIO2XX` rules are partially covered by the port of [flake8_async](https://github.com/astral-sh/ruff/tree/75c9be099f878c381c8734ecd7c1217c44d2b612/crates/ruff_linter/src/rules/flake8_async/rules). 

The recommended fixes in `flake8-trio` are TRIO specific though. How do we want to combine those two?

**Comment by jakkdl** [comment]
whoa, this is super cool!! As the primary author of flake8-trio I'm super honored to be included in ruff :heart: 
Feel free to ping me about any questions of functionality or implementation details
@Zac-HD might also have opinions or suggestions about stuff

**Comment by jakkdl** [comment]
> The recommended fixes in `flake8-trio` are TRIO specific though. How do we want to combine those two?

There's some `anyio` support as well in flake8-trio and for a while we considered renaming the library. If `flake8-trio` sees `import anyio` and no `import trio` in the code, or [--anyio](https://github.com/Zac-HD/flake8-trio#--anyio) is passed it will suggest autofixes from the anyio library instead of trio.

**Comment by jakkdl** [comment]
> TRIO118 sounds like it is not needed in Ruff, similar to TRIO106

TRIO106 definitely isn't needed by itself, and solely a consequence of implementation at the time. TRIO118 is only partially a weakness in implementation, the part about multi-backend programs still holds though.

> The TRIO2XX rules are partially covered by the port of [flake8_async](https://github.com/astral-sh/ruff/tree/75c9be099f878c381c8734ecd7c1217c44d2b612/crates/ruff_linter/src/rules/flake8_async/rules).

Yeah there's overlap there, though I gave the implementation in flake8-trio quite a bit more love and the suggestions/error messages are quite a bit more detailed.



TRIO105 is mostly redundant as of the recently released trio 0.23 which included types into the main package, and type checkers will complain about async functions not being awaited. I suppose it could be useful for somebody not using type-checking, but enabling this check, but I think that's a pretty slim proportion and the rule requires constantly keeping up with the API of anyio/trio.

**Comment by Zac-HD** [comment]
🥳 I'm delighted to see this going into `ruff`!

`TRIO105` doesn't seem redundant to me - many people aren't using type annotations (everywhere), and typecheckers also don't offer autofixers the way that `ruff` can.  Those fixers would mostly be unsafe, but still remarkably useful IMO!

**Comment by charliermarsh** [comment]
Thank you both so much for chiming in here, for all the helpful commentary, and for all your work on flake8-trio (among so much else)!

I already gated one rule behind "require an `import trio` to be present before firing" so we now have some precedent and infrastructure for that, which will likely be helpful with a few of the other rules and their associated fixes.

**Comment by alex007sirois** [comment]
> I already gated one rule behind "require an import trio to be present before firing"

Most of the rules for trio would apply for anyio too:

> > The recommended fixes in `flake8-trio` are TRIO specific though. How do we want to combine those two?
> 
> There's some `anyio` support as well in flake8-trio and for a while we considered renaming the library. If `flake8-trio` sees `import anyio` and no `import trio` in the code, or [--anyio](https://github.com/Zac-HD/flake8-trio#--anyio) is passed it will suggest autofixes from the anyio library instead of trio.

**Comment by jakkdl** [comment]
This issue should be renamed and all the rules in the checklist renamed to match upstream merging with flake8-async. See #10303

Also TRIO117 is removed (because `trio.MultiError` is removed), and there's two new rules ASYNC250 and ASYNC251

**Comment by MichaReiser** [comment]
I updated the issue to reflect the `flake8-trio` to `flake8-async` rename, crossed of the new rules implemented in #10416, and added new rules to the list to match `flake8-async`'s documentation

**Comment by Zac-HD** [comment]
Hey all - I just tried moving some code from `flake8-async` to `ruff check` for that delicious massive speedup, and ended up back on this issue 🙂 

The good news: https://github.com/astral-sh/ruff/pull/10416 also implemented `ASYNC251`!

The sad-for-me news: lots of other rules still to implement.  If you feel inspired but not _so_ inspired as to finish closing out this issue, `ASYNC900` and `ASYNC200` seem pretty simple to implement, and would be _really_ nice to have...

**Comment by Zac-HD** [comment]
Prompted by a conversation elsewhere: if `ruff` starts to handle cross-file analyses, you could avoid `ASYNC114` entirely (ie check whether there's a `task_status=` argument instead of configuring a list) which would be quite nice 🙂

**Comment by charliermarsh** [comment]
Also prompted by a conversation elsewhere: I'd love to see implementations for `ASYNC900`, `ASYNC910`, and `ASYNC911` in particular.

**Comment by jakkdl** [comment]
New rules in upstream https://flake8-async.readthedocs.io/en/latest/rules.html#async121
* `ASYNC121` control-flow-in-taskgroup
* `ASYNC122` delayed-entry-of-relative-cancelscope
* `ASYNC123` bad-exception-group-flattening
  * this one isn't specific to async, and affects anybody trying to extract and reraise an exception contained inside an exception group. Upstream implementation is fairly crude

**Comment by PieterCK** [comment]
Hello folks! Not, sure if this is the best place to ask for reviews, but I've opened #20264 for `ASYNC240`. I'd appreciate it if someone could drop a quick review 🙂

**Comment by LeoVasanko** [comment]
A false alarm with ASYNC240:

```
ASYNC240 Async functions should not use pathlib.Path methods, use trio.Path or anyio.path
  --> jacloud/app.py:47:41
   |
45 |     try:
46 |         await frontend.load()
47 |         await asyncio.to_thread(lambda: plandir.mkdir(parents=True, exist_ok=True))
   |                                         ^^^^^^^^^^^^^
```

It incorrectly considers the lambda - about to be run in a thread - as part of async context of the function we are in. Of course the proper fix is to avoid the unnecessary lambda and this also avoids the ruff error:

```
        await asyncio.to_thread(plandir.mkdir, parents=True, exist_ok=True)
```

**Comment by ntBre** [comment]
@LeoVasanko I'm hopeful that this may have been fixed by https://github.com/astral-sh/ruff/pull/23426, which would have been released in 0.15.2. If this is still happening after 0.15.2, please open a separate issue!

**Comment by SirCypkowskyy** [comment]
I'd like to take a crack at `ASYNC200` (`async200-blocking-calls`) unless someone is already on it. I know the recent maintainer ask in this thread leans toward `ASYNC900`/`ASYNC910`/`ASYNC911`, so this is meant to slot alongside, not in place of, that work.

Intended Ruff config shape:

```toml
[tool.ruff.lint.flake8-async]
async200-blocking-calls = { "requests.*" = "httpx.AsyncClient", "django.db.connection.cursor" = "django.db.aconnection.cursor" }
```

That maps directly onto upstream's `pattern -> replacement` (fnmatch-style patterns, empty default, replacement embedded in the diagnostic message). Scope is just `ASYNC200`; I won't touch `ASYNC900`/`910`/`911` in the same PR.

**Comment by ntBre** [comment]
I don't think we should use `async200` as a prefix on the rule or option names, but that otherwise sounds good to me. Thank you!

**Comment by JoakimUhlin** [comment]
The checklist in the issue be updated to indicate that `ASYNC119`, `ASYNC240` and `ASYNC250` are completed. That is at least what the [official docs claim](https://docs.astral.sh/ruff/rules/#flake8-async-async) (although `ASYNC119` is still in preview).

**Comment by tjkuson** [comment]
Should `ASYNC300` be considered implemented via [`RUF006`](https://docs.astral.sh/ruff/rules/asyncio-dangling-task/)? It's currently unchecked. The only difference I can tell is that the documentation doesn't mention using `asyncio.TaskGroup.create_task` and structured concurrency as an alternative.

**Comment by jakkdl** [comment]
> Should `ASYNC300` be considered implemented via [`RUF006`](https://docs.astral.sh/ruff/rules/asyncio-dangling-task/)? It's currently unchecked. The only difference I can tell is that the documentation doesn't mention using `asyncio.TaskGroup.create_task` and structured concurrency as an alternative.

Yeah they check for the same thing.
`flake8-async` is very opinionated about structured concurrency, but I think an example with using a `TaskGroup` would be nice when it's a relatively new feature in `asyncio`
